app/services/course/user_registration_service.rb

```diff
@@ -0,0 +1,57 @@
+class Course::UserRegistrationService
+  # Registers the specified registration.
+  #
+  # @param [Course::Registration] registration The registration object to be processed.
+  # @return [bool] True if the registration succeeded. False if the registration failed.
+  def register(registration)
+    CourseUser.transaction do
+      if registration.code.empty?
+        register_course_user(registration)
+      else
+        claim_registration_code(registration)
+      end
+    end
+  end
+
+  private
+
+  # Registers the given +user+ for a +course+. This sets the course user to the +requested+ state.
+  #
+  # @param [Course::Registration] registration The registration model containing the course user
+  #   parameters.
+  # @return [bool] True if the creation succeeded.
+  def register_course_user(registration)
+    course_user = CourseUser.new(course: registration.course, user: registration.user,
+                                 creator: registration.user, updater: registration.user)
+    registration.course_user = course_user
+    course_user.save
+  end
+
+  # Claims a given user's registration code. This sets the course user to the +approved+ state.
+  #
+  # @param [Course::Registration] registration The registration model containing the course user
+  #   parameters.
+  # @return [bool] True if the creation succeeded.
+  def claim_registration_code(registration)
```

Comment:
Assignment Branch Condition size for claim_registration_code is too high. [15.17/15]